### ISSUE TITLE
"machinectl clone" chokes when copying a directory tree with AF_UNIX socket nodes

### ISSUE THREAD

**Issue by legrostdg** [systemd/systemd#2914]
### Submission type
- [ ] Bug report
- [X] Request for enhancement (RFE)

_NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!_
### systemd version the issue has been seen with

> 229

_NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!_
### Used distribution

> Debian sid

machinectl should be able to clone chroot directories. Right now:

```
# debootstrap jessie /var/lib/machines/mycontainer
# machinectl list-images
NAME            TYPE      RO  USAGE  CREATED                      MODIFIED
mycontainer     directory no  n/a    n/a                          n/a     
# machinectl clone mycontainer mycontainer2
Could not clone image: Operation not supported
```

Note that the error messages is not very helpful. https://github.com/systemd/systemd/issues/2060

Of course I can do:

```
# machinectl export-tar mycontainer /var/tmp/mycontainer.tar.gz
# rm -rf /var/lib/machines/mycontainer
# machinectl import-tar /var/tmp/mycontainer.tar.gz mycontainer
```

and this new container is clonable

```
# machinectl list-images
NAME            TYPE      RO  USAGE  CREATED                      MODIFIED
mycontainer     subvolume no  60M    Thu 2015-09-24 22:41:39 CEST n/a     
# machinectl clone mycontainer testclone
# machinectl list-images
NAME            TYPE      RO  USAGE  CREATED                      MODIFIED
mycontainer     subvolume no  60M    Thu 2015-09-24 22:41:39 CEST n/a     
testclone       subvolume no  60M    Thu 2015-09-24 22:48:39 CEST n/a     
```

But it is not very user friendly...

This may be related to https://github.com/systemd/systemd/issues/2867 and https://github.com/systemd/systemd/issues/2728

irc ohsix plans to work on this soon.

**Comment by legrostdg** [comment]
Also note that machinectl manpage says:

```
clone NAME NAME
   Clones a container or VM image. The arguments specify the
   name of the image to clone and the name of the newly
   cloned image. Note that plain directory container images
   are cloned into subvolume images with this command.
```

which seems to imply that cloning a chroot should work out of the box.

**Comment by ohsix** [comment]
whoops, sorry if i gave you the impression that i was going to work on it, i just said that it was probably a reasonable RFE

**Comment by poettering** [comment]
@legrostdg so, did i get this right? you have /var/lib/machines on non-btrfs, and a directory tree there you want to clone?

**Comment by legrostdg** [comment]
@poettering I have /var/lib/machines on btrfs, and created a chroot there with debootstrap. Feel free to ask for more information :)

**Comment by poettering** [comment]
hmm, this is very strange. Any chance you could do the following for me: issue "systemctl start start-machined". Then you "strace -s500 -p <pid>" for the PID of systemd-machined, and then run your "machinectl clone" command in a different terminal. THen, switch back, and ctrl-c the strace command. Then, look for EOPNOTSUP erros in that output, and paste the line you find (plus some lines surrounding it).

**Comment by legrostdg** [comment]
I've run the commands you requested. There is no EOPNOTSUP errors in the output...

The error I see is:

```
sendmsg(6, {msg_name(0)=NULL, msg_iov(2)=[{"l\3\1\1\34\0\0\0\27\0\0\0O\0\0\0\5\1u\0\2\0\0\0\6\1s\0\6\0\0\0:1.370\0\0\4\1s\0'\0\0\0org.freedesktop.DBus.Error.NotSupported\0\10\1g\0\1s\0\0", 96}, {"\27\0\0\0Operation not supported\0", 28}], msg_controllen=0, msg_flags=0}, MSG_DONTWAIT|MSG_NOSIGNAL) = 124
```

**Comment by poettering** [comment]
hmm, next step would be to gdb through machined then, and check where the EOPNOTSUPP happens. Do you feel comfortable with gdb? If so, please attach to systemd-machined, set a break point on bus_image_method_clone() and see where this fail. (might make sense to break on image_clone() instead)

**Comment by legrostdg** [comment]
@poettering not so much, I have to admit... But I can try to give you this information with a little more guidance.
Is this setup hard to reproduce? This behaviour seems to be the same on every debian system #2867 #2060

**Comment by poettering** [comment]
@legrostdg hmm, so could you paste the strace output you generated like this? (or at least the last 100 lines or so, given that it might be quite large). THis should give me a hint on what's going on.

I am pretty sure #2867 is a different case than yours though. And in contrast to your case #2060 appears to be a system where /var/lib/machines is not btrfs. Are you sure /var/lib/machines for your is btrfs?

**Comment by legrostdg** [comment]
Here are the last 100 lines of the strace output (when running "machinectl clone android-builder testclone")

```
fcntl(12, F_SETFD, FD_CLOEXEC)          = 0
mkdirat(11, "lib64", 0755)              = 0
openat(11, "lib64", O_RDONLY|O_NOCTTY|O_DIRECTORY|O_NOFOLLOW|O_CLOEXEC) = 13
fchown(13, 0, 0)                        = 0
fchmod(13, 0755)                        = 0
utimensat(13, NULL, [{1443110144, 590533072}, {1449383742, 380526438}], 0) = 0
flistxattr(12, "", 100)                 = 0
getdents(12, /* 3 entries */, 32768)    = 88
newfstatat(12, "ld-linux-x86-64.so.2", {st_mode=S_IFLNK|0777, st_size=32, ...}, AT_SYMLINK_NOFOLLOW) = 0
readlinkat(12, "ld-linux-x86-64.so.2", "/lib/x86_64-linux-gnu/ld-2.21.so", 99) = 32
symlinkat("/lib/x86_64-linux-gnu/ld-2.21.so", 13, "ld-linux-x86-64.so.2") = 0
fchownat(13, "ld-linux-x86-64.so.2", 0, 0, AT_SYMLINK_NOFOLLOW) = 0
getdents(12, /* 0 entries */, 32768)    = 0
close(12)                               = 0
close(13)                               = 0
newfstatat(10, "mnt", {st_mode=S_IFDIR|0755, st_size=0, ...}, AT_SYMLINK_NOFOLLOW) = 0
openat(10, "mnt", O_RDONLY|O_NOCTTY|O_DIRECTORY|O_NOFOLLOW|O_CLOEXEC) = 12
fstat(12, {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
fcntl(12, F_GETFL)                      = 0x38000 (flags O_RDONLY|O_LARGEFILE|O_DIRECTORY|O_NOFOLLOW)
fcntl(12, F_SETFD, FD_CLOEXEC)          = 0
mkdirat(11, "mnt", 0755)                = 0
openat(11, "mnt", O_RDONLY|O_NOCTTY|O_DIRECTORY|O_NOFOLLOW|O_CLOEXEC) = 13
fchown(13, 0, 0)                        = 0
fchmod(13, 0755)                        = 0
utimensat(13, NULL, [{1443110147, 502539132}, {1443110147, 502539132}], 0) = 0
flistxattr(12, "", 100)                 = 0
getdents(12, /* 2 entries */, 32768)    = 48
getdents(12, /* 0 entries */, 32768)    = 0
close(12)                               = 0
close(13)                               = 0
newfstatat(10, "srv", {st_mode=S_IFDIR|0755, st_size=0, ...}, AT_SYMLINK_NOFOLLOW) = 0
openat(10, "srv", O_RDONLY|O_NOCTTY|O_DIRECTORY|O_NOFOLLOW|O_CLOEXEC) = 12
fstat(12, {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
fcntl(12, F_GETFL)                      = 0x38000 (flags O_RDONLY|O_LARGEFILE|O_DIRECTORY|O_NOFOLLOW)
fcntl(12, F_SETFD, FD_CLOEXEC)          = 0
mkdirat(11, "srv", 0755)                = 0
openat(11, "srv", O_RDONLY|O_NOCTTY|O_DIRECTORY|O_NOFOLLOW|O_CLOEXEC) = 13
fchown(13, 0, 0)                        = 0
fchmod(13, 0755)                        = 0
utimensat(13, NULL, [{1443110147, 506539140}, {1443110147, 506539140}], 0) = 0
flistxattr(12, "", 100)                 = 0
getdents(12, /* 2 entries */, 32768)    = 48
getdents(12, /* 0 entries */, 32768)    = 0
close(12)                               = 0
close(13)                               = 0
newfstatat(10, "opt", {st_mode=S_IFDIR|0755, st_size=0, ...}, AT_SYMLINK_NOFOLLOW) = 0
openat(10, "opt", O_RDONLY|O_NOCTTY|O_DIRECTORY|O_NOFOLLOW|O_CLOEXEC) = 12
fstat(12, {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
fcntl(12, F_GETFL)                      = 0x38000 (flags O_RDONLY|O_LARGEFILE|O_DIRECTORY|O_NOFOLLOW)
fcntl(12, F_SETFD, FD_CLOEXEC)          = 0
mkdirat(11, "opt", 0755)                = 0
openat(11, "opt", O_RDONLY|O_NOCTTY|O_DIRECTORY|O_NOFOLLOW|O_CLOEXEC) = 13
fchown(13, 0, 0)                        = 0
fchmod(13, 0755)                        = 0
utimensat(13, NULL, [{1443110147, 514539157}, {1443110147, 514539157}], 0) = 0
flistxattr(12, "", 100)                 = 0
getdents(12, /* 2 entries */, 32768)    = 48
getdents(12, /* 0 entries */, 32768)    = 0
close(12)                               = 0
close(13)                               = 0
newfstatat(10, "media", {st_mode=S_IFDIR|0755, st_size=0, ...}, AT_SYMLINK_NOFOLLOW) = 0
openat(10, "media", O_RDONLY|O_NOCTTY|O_DIRECTORY|O_NOFOLLOW|O_CLOEXEC) = 12
fstat(12, {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
fcntl(12, F_GETFL)                      = 0x38000 (flags O_RDONLY|O_LARGEFILE|O_DIRECTORY|O_NOFOLLOW)
fcntl(12, F_SETFD, FD_CLOEXEC)          = 0
mkdirat(11, "media", 0755)              = 0
openat(11, "media", O_RDONLY|O_NOCTTY|O_DIRECTORY|O_NOFOLLOW|O_CLOEXEC) = 13
fchown(13, 0, 0)                        = 0
fchmod(13, 0755)                        = 0
utimensat(13, NULL, [{1443110147, 526539182}, {1443110147, 526539182}], 0) = 0
flistxattr(12, "", 100)                 = 0
getdents(12, /* 2 entries */, 32768)    = 48
getdents(12, /* 0 entries */, 32768)    = 0
close(12)                               = 0
close(13)                               = 0
getdents(10, /* 0 entries */, 32768)    = 0
brk(0x55732cc99000)                     = 0x55732cc99000
close(10)                               = 0
close(11)                               = 0
open("/var/lib/machines", O_RDONLY|O_NOCTTY|O_DIRECTORY|O_CLOEXEC) = 10
fstat(10, {st_mode=S_IFDIR|0700, st_size=42, ...}) = 0
openat(10, "testclone", O_RDONLY|O_NOCTTY|O_DIRECTORY|O_CLOEXEC) = 11
fstatfs(11, {f_type="BTRFS_SUPER_MAGIC", f_bsize=4096, f_blocks=56209745, f_bfree=10368577, f_bavail=8958456, f_files=0, f_ffree=0, f_fsid={-739180754, -2139978435}, f_namelen=255, f_frsize=4096, f_flags=1056}) = 0
ioctl(11, BTRFS_IOC_INO_LOOKUP, 0x7ffc64af6140) = 0
ioctl(10, BTRFS_IOC_SNAP_DESTROY, 0x7ffc64af7240) = 0
fstatfs(10, {f_type="BTRFS_SUPER_MAGIC", f_bsize=4096, f_blocks=56209745, f_bfree=10368577, f_bavail=8958456, f_files=0, f_ffree=0, f_fsid={-739180754, -2139978214}, f_namelen=255, f_frsize=4096, f_flags=1056}) = 0
ioctl(10, BTRFS_IOC_TREE_SEARCH, 0x7ffc64af60c0) = 0
fstatfs(10, {f_type="BTRFS_SUPER_MAGIC", f_bsize=4096, f_blocks=56209745, f_bfree=10368577, f_bavail=8958456, f_files=0, f_ffree=0, f_fsid={-739180754, -2139978214}, f_namelen=255, f_frsize=4096, f_flags=1056}) = 0
ioctl(10, BTRFS_IOC_QGROUP_CREATE, 0x7ffc64af70d0) = 0
close(11)                               = 0
close(10)                               = 0
close(9)                                = 0
open("/var/lib/container/android-builder", O_RDONLY|O_NOCTTY|O_DIRECTORY|O_CLOEXEC) = 9
fstat(9, {st_mode=S_IFDIR|0755, st_size=132, ...}) = 0
close(9)                                = 0
unlink("/run/systemd/nspawn/locks/name-testclone") = 0
close(8)                                = 0
sendmsg(6, {msg_name(0)=NULL, msg_iov(2)=[{"l\3\1\1\34\0\0\0\30\0\0\0O\0\0\0\5\1u\0\2\0\0\0\6\1s\0\6\0\0\0:1.243\0\0\4\1s\0'\0\0\0org.freedesktop.DBus.Error.NotSupported\0\10\1g\0\1s\0\0", 96}, {"\27\0\0\0Operation not supported\0", 28}], msg_controllen=0, msg_flags=0}, MSG_DONTWAIT|MSG_NOSIGNAL) = 124
epoll_wait(4, strace: Process 382 detached
 <detached ...>
```

**Comment by poettering** [comment]
@legrostdg ok, that was very useful, but it appears the issue is triggered earlier than the strace shows. Could you upload the full strace for the clone operation somewhere please?

**Comment by poettering** [comment]
@legrostdg is it possible that your directory tree already contains submounts or so? or is it just a naked btrfs directory with nothing weird at all?

**Comment by legrostdg** [comment]
I have one btrfs volume and I use subvolumes for /, /home, /var, /usr

**Comment by legrostdg** [comment]
[strace.txt](https://github.com/systemd/systemd/files/242331/strace.txt)

**Comment by poettering** [comment]
OK, I think I figured it out. Our copy routine can't copy UNIX sockets, and returns EOPNOTSUPP in that case. Should be fixable.

Thanks for providing the strace output.

**Comment by poettering** [comment]
Fix waiting in #3149.

**Comment by legrostdg** [comment]
Great! Thanks!
